jupyter_server_proxy/handlers.py

```diff
@@ -213,7 +213,18 @@ def _check_host_whitelist(self, host):
         client = httpclient.AsyncHTTPClient()
 
         req = self._build_proxy_request(host, port, proxied_path, body)
-        response = await client.fetch(req, raise_error=False)
+
+        try:
+            response = await client.fetch(req, raise_error=False)
+        except httpclient.HTTPError as err:
```

Comment:
The `client.fetch` call uses `raise_error=False`. I think this means no exceptions will be raised. Is this correct or is the argument confusingly named (no exceptions except for timeouts)?

If there are no exceptions we should probably deal with timeouts in the code below

Reply:
We do see the timeout exception even with `raise_error=False`, see the stack trace here: https://github.com/jupyterhub/jupyter-server-proxy/issues/182.

In the [docs](https://www.tornadoweb.org/en/stable/httpclient.html#tornado.httpclient.AsyncHTTPClient.fetch), it said explicitly it will not suppress all errors:
```
The raise_error=False argument only affects the HTTPError raised when a non-200 response code is used, instead of suppressing all errors.
```

So why the timeout error is not `HTTPError` and being suppressed? I guess if there is a timeout, then there won't be any response, thus the error cannot be suppressed, I think.

Reply:
Aha! Thanks for digging out the docs :)

Maybe we can add a comment to the code which contains your response/link to docs to remind future generations that this seemingly weird setup is on purpose.

Reply:
Good idea, done!

Reply:
@betatim I think raise_error=False tells the client to return the response to the caller, regardless of error code, rather than raise an exception. I think we want to always get the response, but we could conceivably return other things to the user based on the http code if we wanted.

Also, in tornado <= 5.1 it would [set the return code for errors to 599](https://www.tornadoweb.org/en/branch5.1/httpclient.html#tornado.httpclient.AsyncHTTPClient.fetch). If people are still using 5.1 with this proxy and this PR, they may start to see a lot of timeout errors whereas before we would just return the response without checking the error code. I don't think we care too much but it's worth noting. If a lot of people are still on 5.1, I think we'd have to insert a check on the tornado version and use the old behavior for <= 5.1.

Reply:
@ryanlovett Could you please clarify why `they may start to see a lot of timeout errors whereas before we would just return the response without checking the error code.`?  

To be clear, the only thing we did in this PR is to capture the error which previously treated as internal server error (we are not changing the setting of `raise_error=False`). In the link you mentioned, it said:
> The raise_error=False argument currently suppresses all errors, encapsulating them in HTTPResponse objects with a 599 response code. This will change in Tornado 6.0: raise_error=False will only affect the HTTPError raised when a non-200 response code is used.

If I understand correctly, tornado <=5.1 won't raise any error, so the code should function exactly as before. And for tornado >5.1 (I am testing with 6.0+), it will raise the error then we will start to return timeout errors instead of saying the internal server error.

Reply:
We won't see more exceptions due to raise_error=False, but when suppressing the exceptions, tornado <= 5.1 would set the return code to be 599. This PR examines that code... only when there is an exception. So... never mind. Sorry.

Reply:
This is exciting guys!  Can't wait to try this out!  Thanks @oeway!

Reply:
@betatim Any final thoughts before merging?

Reply:
No thoughts. My confusion was from the "split" behaviour of the new meaning of the argument. That was addressed in https://github.com/jupyterhub/jupyter-server-proxy/pull/183#discussion_r398504580. Marking this conversation as resolved.